src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php

```diff
@@ -0,0 +1,260 @@
+            $builder
+        );
+    }
+
+    /**
+     * It should support Not.
+     */
+    public function testNot()
+    {
+        $builder = $this->manager->createQueryBuilder();
+        $builder->setLocale('de')
+            ->from()->document('page', 'p')->end()
+            ->where()->not()->eq()->field('p.redirectType')->literal('hello');
+
+        $this->assertQuery(
+            'SELECT * FROM [nt:unstructured] AS p WHERE ((NOT p.[i18n:de-nodeType] = \'hello\') AND p.[jcr:mixinTypes] = \'sulu:page\')',
+            $builder
+        );
+    }
+
+    /**
+     * It should maintain the Sulu Wrapper.
+     */
+    public function testMaintainWrapper()
+    {
+        $builder = $this->manager->createQueryBuilder();
+        $where = $builder->setLocale('de')
+            ->useStructure('p', 'overview')
+            ->from()->document('page', 'p')->end()
+            ->where()
+                ->andX()
+                    ->orX()
+                        ->eq()->field('p.redirectType')->literal('hello')->end()
+                        ->eq()->structureField('p.article')->literal('bar')->end()
+                    ->end()
+                    ->eq()->structureField('p.article')->literal('barbar')
+                ->end()
+            ->end();
+
+        $this->assertInstanceOf('Sulu\Component\Content\Document\Query\Builder\Where', $where);
+    }
+
+    /**
+     * It should support ordering.
+     */
+    public function testOrdering()
+    {
+        $builder = $this->manager->createQueryBuilder();
+        $builder->setLocale('de')
+            ->useStructure('p', 'overview')
+            ->from()->document('page', 'p')->end()
+            ->orderBy()
+                ->asc()->field('p.redirectType')->end()
+                ->desc()->structureField('p.article');
+
+        $this->assertQuery(
+            'SELECT * FROM [nt:unstructured] AS p WHERE p.[jcr:mixinTypes] = \'sulu:page\' ORDER BY p.[i18n:de-nodeType] ASC, p.[i18n:de-article] DESC',
+            $builder
+        );
+    }
+
+    /**
+     * It should support ordering.
+     */
+    public function testOrderAdd()
+    {
+        $builder = $this->manager->createQueryBuilder();
+        $builder->setLocale('de')
+            ->useStructure('p', 'overview')
+            ->from()->document('page', 'p')->end()
+            ->orderBy()
+                ->asc()->field('p.redirectType')->end()
+            ->end()
+            ->addOrderBy()
+                ->desc()->structureField('p.article');
+
+        $this->assertQuery(
+            'SELECT * FROM [nt:unstructured] AS p WHERE p.[jcr:mixinTypes] = \'sulu:page\' ORDER BY p.[i18n:de-nodeType] ASC, p.[i18n:de-article] DESC',
+            $builder
+        );
+    }
+
+    /**
+     * It should throw an exception if an unknown structure field is used.
+     *
+     * @expectedException InvalidArgumentException 
```

Comment:
PHPUnit will always evaluate this absolutely. It is not aware of
namespaces here.

On Tue, Feb 16, 2016 at 11:23:20PM -0800, Daniel Rotter wrote:
>    In
>    [1]src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php:
> 
>  > +            ->orderBy()
>  > +                ->asc()->field('p.redirectType')->end()
>  > +            ->end()
>  > +            ->addOrderBy()
>  > +                ->desc()->structureField('p.article');
>  > +
>  > +        $this->assertQuery(
>  > +            'SELECT * FROM [nt:unstructured] AS p WHERE p.[jcr:mixinTypes] = \'sulu:page\' ORDER BY p.[i18n:de-nodeType] ASC, p.[i18n:de-article] DESC',
>  > +            $builder
>  > +        );
>  > +    }
>  > +
>  > +    /**
>  > +     * It should throw an exception if an unknown structure field is used.
>  > +     *
>  > +     * @expectedException InvalidArgumentException
> 
>    Btw, I think it's marked as wrong because it might be ambiguous. If there
>    is another InvalidArgumentException added to this namespace later, this
>    might lead to wrong results.
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/sulu-io/sulu/pull/1863#discussion_r53128345
>    2. https://github.com/sulu-io/sulu/pull/1863/files#r53128345
>    3. https://github.com/sulu-io/sulu/pull/1863/files#r53128345